clarity-ext-scripts/clarity_ext_scripts/covid/fetch_biobank_barcodes.py

```diff
@@ -0,0 +1,137 @@
+        filenames = self.context.file_service.list_filenames('Raw sample list')
+        base_names = [n.split('.')[0] for n in filenames]
+        plate_barcode = single(list(plate_barcodes))
+        if plate_barcode not in base_names:
+            raise UsageError(
+                "The 'Raw sample list' name is not matching with the plate "
+                "barcode in 'Raw biobank list', {}".format(plate_barcode))
+
+        # Validate that all wells in biobank file are matched in sample list file
+        file_stream2 = self.context.local_shared_file('Raw sample list')
+        sample_matrix = self._build_sample_matrix(file_stream2, plate_barcode)
+        for key in biobank_matrix:
+            try:
+                if biobank_matrix[key]['biobank_barcode'] != 'NO TUBE':
+                    _ = sample_matrix[key]['Sample Id']
+            except KeyError:
+                raise UsageError(
+                    "The well references are not matching between the "
+                    "files in 'Raw biobank list' and 'Raw sample list, "
+                    "well '{}' has no match in the 'Raw sample list'"
+                    .format(biobank_matrix[key]['well'])
+                )
+
+        # Validate that 'NO TUBE' entries in biobank file is empty in sample list
+        # Perhaps this is a little too aggressive validation?
+        sample_matrix_keys = [k for k in sample_matrix]
+        for key in biobank_matrix:
+            if biobank_matrix[key]['biobank_barcode'] == 'NO TUBE' \
+                    and key in sample_matrix_keys \
+                    and sample_matrix[key]['Sample Id']:
+                biobank_well = biobank_matrix[key]['well']
+                sample_list_well = sample_matrix[key]['Position']
+                raise UsageError(
+                    "There is an empty entry in the biobank barcode file "
+                    "that is not empty in the sample list file, "
+                    "biobank well: {}, sample list well: {}"
+                    .format(biobank_well, sample_list_well))
+
+    def biobank_barcode_by_sample_referal_code(self):
+        file_stream = self.context.local_shared_file('Raw biobank list')
+        biobank_matrix = self._build_biobank_matrix(file_stream)
+        plate_barcode = self._plate_barcode_from(biobank_matrix)
+        file_stream2 = self.context.local_shared_file('Raw sample list')
+        sample_matrix = self._build_sample_matrix(file_stream2, plate_barcode)
+        barcode_map = dict()
+        for key in biobank_matrix:
+            if biobank_matrix[key]['biobank_barcode'] == 'NO TUBE':
+                continue
+            biobank_barcode = biobank_matrix[key]['biobank_barcode']
+            sample_referal_code = sample_matrix[key]['Sample Id']
+            barcode_map[sample_referal_code] = biobank_barcode
+
+        return barcode_map
+
+    def _build_sample_matrix(self, file_stream, plate_barcode):
+        csv = Csv(file_stream)
+        matrix = dict()
```

Comment:
Similar to that other matrix name, I would personally prefer `sample_info_by_well_barcode` or something similar.